ui/library/src/utils/VisibilityDetector.ts

```diff
@@ -0,0 +1,69 @@
+export type SetEntriesCallback<TEntry> = (visible: Set<TEntry>) => void;
+export type OnVisibilityChangeCallback<TEntry> = (args: {
+  visibleElements: Element[];
+  hiddenElements: Element[];
+  lastEntries: Set<TEntry>;
+}) => Set<TEntry>;
+
+export default class VisibilityDetector<TEntry> {
+  _root: Element;
+  _observer: IntersectionObserver;
+  _lastEntries: Set<TEntry>;
+  _setEntries: SetEntriesCallback<TEntry>;
+  _onVisibilityChange: OnVisibilityChangeCallback<TEntry>;
```

Comment:
The name `lastEntries` and `setEntries` are a bit general to me. It would be great if you can modify them as below:
```suggestion
  _lastVisibleEntries: Set<TEntry>;
  _setVisibleEntries: SetVisibleEntriesCallback<TEntry>;
  _onVisibleEntriesChange: OnVisibleEntriesChangeCallback<TEntry>;
```